### PR TITLE
Vidur/updating ros unity tutorials

### PR DESCRIPTION
**Pull request by vidurvij-Unity**
## Proposed change(s)

Adds a section in publisher and subscriber tutorials describing the ROSSubscriber and ROSPublisher  example scripts and the  APIs used in them.

### Useful links (GitHub issues, JIRA tickets, forum threads, etc.)

Provide any relevant links here.

https://jira.unity3d.com/browse/AIRO-593
https://github.com/Unity-Technologies/Unity-Robotics-Hub/issues/196
### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [x] Documentation update
- [ ] Other (please describe)

## Testing and Verification

Did a spellcheck and grammar check.

### Test Configuration:
- Unity Version: [e.g. Unity 2020.2.0f1]
- Unity machine OS + version: [e.g. Windows 10]
- ROS machine OS + version: [e.g. Ubuntu 18.04, ROS Noetic]
- ROS–Unity communication: [e.g. Docker]

## Checklist
- [x] Ensured this PR is up-to-date with the `dev` branch
- [x] Created this PR to target the `dev` branch
- [ ] Followed the style guidelines as described in the [Contribution Guidelines](../CONTRIBUTING.md)
- [ ] Added tests that prove my fix is effective or that my feature works
- [x] Updated the documentation as appropriate

## Other comments

### CODE REVIEW

File: tutorials/ros_unity_integration/publisher.md [hunk 1]
```diff
@@ -103,4 +103,120 @@ public class RosPublisherExample : MonoBehaviour
...
+    void Start()
+    {
+        // start the ROS connection
+        ros = ROSConnection.instance;
+    }
+
+    private void Update()
+    {
+        timeElapsed += Time.deltaTime;
+
```
Comment (hyounesy) [inline tutorials/ros_unity_integration/publisher.md:138]: This entire code block was previously shown. what is the purpose for repeating it?

File: tutorials/ros_unity_integration/publisher.md [hunk 2]
```diff
@@ -103,4 +103,120 @@ public class RosPublisherExample : MonoBehaviour
 
 ![](images/tcp_1.gif)
 
+## ROSSPublisherExample script
```
Comment (hyounesy) [inline tutorials/ros_unity_integration/publisher.md:106]: typo: ROS~~S~~PublisherExample

File: tutorials/ros_unity_integration/publisher.md [hunk 3]
```diff
@@ -103,4 +103,120 @@ public class RosPublisherExample : MonoBehaviour
...
+
+### Instantiating the ROSConnection class in unity
+
+```csharp
+void Start()
+    {
+        // start the ROS connection
+        ros = ROSConnection.instance;
+    }
+```
```
Comment (hyounesy) [inline tutorials/ros_unity_integration/publisher.md:178]: inconsistent indentation

File: tutorials/ros_unity_integration/publisher.md [hunk 4]
```diff
@@ -103,4 +103,120 @@ public class RosPublisherExample : MonoBehaviour
...
+
+            MPosRot cubePos = new MPosRot(
+                cube.transform.position.x,
+                cube.transform.position.y,
+                cube.transform.position.z,
+                cube.transform.rotation.x,
+                cube.transform.rotation.y,
+                cube.transform.rotation.z,
+                cube.transform.rotation.w
+            );
```
Comment (hyounesy) [inline tutorials/ros_unity_integration/publisher.md:198]: inconsistent indentation

File: tutorials/ros_unity_integration/subscriber.md [hunk 5]
```diff
@@ -65,4 +65,76 @@ public class RosSubscriberExample : MonoBehaviour
...
+using RosColor = RosMessageTypes.RoboticsDemo.MUnityColor;
+```
+
+### Instantiating the ROS Subscriber in unity
+
+```csharp
+void Start()
+    {
+        ROSConnection.instance.Subscribe<RosColor>("color", ColorChange);
+    }
```
Comment (hyounesy) [inline tutorials/ros_unity_integration/subscriber.md:105]: inconsistent indentation

File: tutorials/ros_unity_integration/subscriber.md [hunk 6]
```diff
@@ -65,4 +65,76 @@ public class RosSubscriberExample : MonoBehaviour
...
+| `string topic`  | Name of the topic to be subscribed on the ROS machine  |
+| `Action<T> callback`  | Callback function used to process the received message |
+
+### Implementing the callback function
+
+```csharp
+void ColorChange(RosColor colorMessage)
+    {
+        cube.GetComponent<Renderer>().material.color = new Color32((byte)colorMessage.r, (byte)colorMessage.g, (byte)colorMessage.b, (byte)colorMessage.a);
+    }
```
Comment (hyounesy) [inline tutorials/ros_unity_integration/subscriber.md:129]: inconsistent indentation

File: tutorials/ros_unity_integration/subscriber.md [hunk 7]
```diff
@@ -65,4 +65,76 @@ public class RosSubscriberExample : MonoBehaviour
...
+We declare the topic to be subscribed before the program execution starts. This is done in the `Start()` function of the MonoBehavior. We add the topic we want to subscribe to the internal dictionary of our ROSConnection instance using `Subscribe` API.
+
+### Subscribe API
+```csharp
+public void Subscribe<T>(string topic, Action<T> callback)
+```
+### Summary
+Its a non-static member of ROSConnection class. It is used to add a ROS subscriber to the internal ROSConnection Dictionary.
+### Type Parameter
+T: ROS message type published in the topic to be subscribed.
```
Comment (hyounesy) [inline tutorials/ros_unity_integration/subscriber.md:116]: Can this sentence be simplified to read better. It took me some effort to parse it in my head. something like:
"Type of the ROS message that we are subscribing to."

File: tutorials/ros_unity_integration/subscriber.md [hunk 8]
```diff
@@ -65,4 +65,76 @@ public class RosSubscriberExample : MonoBehaviour
...
+### Implementing the callback function
+
+```csharp
+void ColorChange(RosColor colorMessage)
+    {
+        cube.GetComponent<Renderer>().material.color = new Color32((byte)colorMessage.r, (byte)colorMessage.g, (byte)colorMessage.b, (byte)colorMessage.a);
+    }
+```
+
+After a message is received in Unity, a callback function is called to process the incoming message. The callback is a delegate which defines a function that receives one parameter and returns nothing as described [here](https://docs.microsoft.com/en-us/dotnet/api/system.action-1?view=net-5.0). In this Subscribe API, that parameter is the message published in the topic.
```
Comment (hyounesy) [inline tutorials/ros_unity_integration/subscriber.md:132]: "published ~~in~~ [to] the topic"

File: tutorials/ros_unity_integration/publisher.md [hunk 9]
```diff
@@ -103,4 +103,120 @@ public class RosPublisherExample : MonoBehaviour
...
+ros.Send(topicName, cubePos);
+```
+After we have created the message, we use the `Send` API to publish the topic.
+
+### Send API
+```csharp
+public async void Send(string rosTopicName, Message message)
+```
+### Summary
+Its a non-static member of ROSConnection class. Its used to send a ROS message to the desired topic on the ROS machine.
```
Comment (hyounesy) [inline tutorials/ros_unity_integration/publisher.md:214]: It~~s~~ [is] used to send a ROS message to the ~~desired~~ [specified] topic ...

**Review by hyounesy — CHANGES_REQUESTED** [review]
- [ ] Remove the dublicate ROSPublisher code block.
 - [ ] Fix inconsistent indentations
 - [ ] Fix typos

File: tutorials/ros_unity_integration/publisher.md [hunk 10]
```diff
@@ -103,4 +103,120 @@ public class RosPublisherExample : MonoBehaviour
...
+
+    private void Update()
+    {
+        timeElapsed += Time.deltaTime;
+
+        if (timeElapsed > publishMessageFrequency)
+        {
+            cube.transform.rotation = Random.rotation;
+
+            MPosRot cubePos = new MPosRot(
```
Comment (LaurieCheers-unity) [inline tutorials/ros_unity_integration/publisher.md:143]: This is old code, need to merge the latest changes to dev branch

**Comment by sdiao** [comment]
@vidurvij-Unity Is this PR still relevant? Should we close it without merging?

**Comment by vidurvij-Unity** [comment]
The tutorial update is deprecated.

**Pull request closed without merging**